Multiple buy menu tweaks

MaxShadow09 opened the pull request:
Changes:
- Pressing "next/prev body" switches to the next or previous category respectively
- Initially, only Crafts, Bodies and Sets tabs are visible. When the user selects an AHuman body, it goes to "equipment selection" mode
- While in equipment selection mode, only Tools, Guns, Bombs and Shields tabs are visible
- While in equipment selection mode, pressing "Back" goes back to the main buy menu screen
- Deleting an AHuman body also deletes all the items after it
- The scroll wheel scrolls the item list, instead of working like "Up/Down" buttons
- The reload button confirms the purchase without having to navigate to the "Buy" button

The idea is to make the Buy Menu a bit more practical, using both mouse and keyboard to quickly add items to the cart. When the user selects a humanoid body, there's a high chance he'll want to add equipment to it, so the menu now displays equipment tabs right after selecting a body.

I still haven't figured out a good way to add loose items. It's possible, but not very practical. The user needs to clear the cart, add a random AHuman body, delete it while he's in equipment selection mode, and then select the items. I will probably need to add an extra button to open equipment selection mode without a body, allowing the user to add items to the top of the list directly.

garethyr commented:
Hey MaxShadow, sorry it took so long for me to get to this. Code comments will come later, but for now I wanted to share my impressions and thoughts.
I like a lot of what you did, it's kind of like work I did back when cc first went open source, except you actually finished it and added some extra stuff to it. It'd definitely serve as a great stopgap to a full redesign of the buy menu!

That said, there are some issues/questions:

1. As you say, we need a way to buy loose equipment or bombs but with things as you have them, I'm not sure what the best way to do that would be. However, thinking on it, I believe hiding menu options isn't actually worth it, and getting rid of that solves the problem. As I see it, the real benefits here are the automatic transition when you select an actor, and going back to actors when you right click. Losing the not-currently-relevant tabs is visually noticeable, but doesn't really affect the actual ux of it, at least with a mouse (maybe it matters more for keyboard-only/controller players). Also, with the addition of q + e swapping tabs, I'd say it matters even less, cause you can swap through things super quick.

2. I'm not sure about the mouse wheel scroll thing. I actually prefer this personally, but I think some people might prefer the old item-by-item version, cause it gives you more careful control.

3. While I like the quick swapping tabs, I wonder if some people will be upset at losing the ability to swap between actors while in the buy menu? Maybe shift or ctrl + mousewheel could be an alternate way of doing this? Though maybe this is better anyway, I suppose this is something other folk should weigh in on.

4. This one's important, have you tried these changes out with controllers or keyboard-only to see how things feel and be sure they work well? Aside from everything else here, someone should test it on every control scheme before it gets merged in.

MaxShadow09 commented:
Hi @garethyr , thanks for reviewing this! I'll answer the questions and issues you described above.
  1. Like you said, the main purpose of the equipment selection mode was to allow the user to quickly select a weapon, and then go back to where he was with a single button. If I left all the tabs visible, it wouldn't be clear when you are in equipment selection mode. I could remove the equipment selection mode altogether (so it only switches to the guns tab when selecting an actor), but then the user would need to switch back to actors manually, giving the impression that selecting an actor triggers a one-way action. I thought about two alternatives for this:
     * (1) I could add an extra button. It would look like another tab (after the last one), but with a different color, and the label "Equip.". Clicking it would enable equipment selection mode manually, so the user can add loose items. I could also change the buy menu background color when equipment selection is enabled (the same color as the Equip. button) to make it more intuitive. I still need to figure out how this would work when switching tabs using keys or controller buttons, because that opens the next tab automatically, and this new Equip. button is not really a tab. Maybe it's more intuitive if I make it look like the preset buttons instead of a tab.
     * (2) I could disable the main buttons (Crafts, actors, presets) while equipment selection is enabled, without hiding them. When the user selects an actor, those 3 buttons would look greyed out/transparent to indicate they are temporarily disabled. However, when equipment selection is disabled, all buttons are visible and enabled, so the user can add any kind of item to the list. This way all the current controls remain untouched, while still being able to add loose items.
     * In both cases, selecting any item when not in equipment selection mode would add them to the top of the list (so they are always loose items, even if the user added some actors first).
  2. I never liked the original scroll. The new one allows me to find items much faster. While I understand some people may want to use the old one, that could apply to all the changes I made. Maybe I could add checkboxes to the options menu to enable all these features individually.
  3. I replaced those two buttons because I never used them to switch actors while the buy menu was open. Usually, I try to spend as little time as possible in the buy menu, so I focus on filling the cart quickly, not paying much attention to what's happening on the battlefield. To make all these features work nicely with controllers, I tried to use only buttons that can be mapped in the options menu (I can't remember if CTRL does anything with the default controls). Maybe I could add an extra mappable input called "Action modifier" or something like that. While holding that button, the "next/prev body" buttons actions change, although I'm not sure if switching actors should be the default action or the alternative one.
  4. I couldn't test this properly with a controller because I'm using a generic directInput controller virtualized to an Xbox360 controller. It doesn't work well in CCCP (although it works well in the Steam version). However, all the new controls are based on buttons that can be mapped in the options menu, so it should work fine except for the scroll. The buy menu navigation is also the same for both controller and keyboard users, and in my tests it worked fine.

Like I said some time ago, this was mostly an experiment I did for myself, so some changes are a bit drastic and may not be everyone's cup of tea. And most things I changed were pretty basic. I could implement these changes if necessary, but it may take some time, specially if they involve changing the default controls or altering the UI.

garethyr commented:
I think for this set of work, it would be good to keep it simple. I'm not saying you have to do this or anything, but I think the buy menu needs a total ground-up redesign (see [here](https://github.com/cortex-command-community/Cortex-Command-Community-Project-Data/wiki/Buy-Menu-Revamp), and feel free to share your thoughts), so it doesn't make a ton of sense to go too hardcore on something that's gonna be redone. However, as I said before, I like a lot of the little tweaks you put together.
But yeah, I think that the concept of equipment mode is a little overkill; the real helpful thing for end users is the quick actions - adding an actor taking you to weapon selection, and pressing the pie menu button once you're there taking you back to actor selection. The hidden or disabled tabs don't actually add much to this, and they cause problems with loose items, as we've discussed. So it seems to me that simplifying things so you get rid of the equipment mode concept but keep the nice quick actions would be a good middleground.

Scrolling - fair enough, I prefer this too. Maybe we just go with it and, then put a toggle for it if people complain. If you wanna not worry about adding an option in the settings menu, you can just put it as a settings.ini flag and we can tell people to edit it manually. Not the most elegant option, but it does the trick.

Actor selection - again, fair enough. I'm personally fine with it, I think this is more useful than selecting actors from the buy menu anyway.

Controller - hmmm, was it working through steam big picture mode or whatever? Steam has its own built-in controller stuff that tended to conflict with cc's. If you download B33 from [here ](https://discord.com/channels/493464078417657866/493465722320715787/744943049078997073) (which is the same as the last steam beta), I'm curious if it'll be better than cccp.

And yeah, I totally get it that this was just you trying things out. Thanks for putting together a PR with your work and taking the time to discuss it, I appreciate it :)

traunts commented:
My input on this:
- Scroll change is welcome, feels far more useful.
- Use of the prev/next buttons to change category is also excellent, but should loop.
- Reload to confirm isn't useful for mouse+keyboard as you can just hit m1 with the mouse cursor outside the GUI to do the same thing, reasonable for Keyboard though again, same can be achieved by spamming right and then fire. No reason not to implement this all the same.
- If using a mouse+keyboard, and the 'back' button is clicked while the cursor is outside the GUI, it should close the menu not just return to body selection. This was a personal note, where it felt difficult when I wanted to rapidly exit, and would align with 'clicking outside the gui to either close or confirm the order'.
- Finally, I agree that it shouldn't hide tabs. With the quickactions, and the auto switching, the easiest solution is just to have them all visible, otherwise I hit really annoying situations where, if I'd accidentally added an actor who I wanted to equip, or decided that one needed another weapon or tool having gone back, there was no easy way to add them. As Gacyr said, this would also allow for the whole loose items thing. 

That's about my thoughts on the matter. Good, simple changes, though not too many that they'd be wasted when it's reworked in the future.

MaxShadow09 commented:
Thanks for the feedback @traunts . I'll see if I can find a way to keep the whole "Inventory selection mode" without hiding the tabs. I already have something in mind. I also agree the next/prev actor keys should loop the selected tabs, I'll change that. About closing the tab with the back button, I'm not sure if that could be a bit confusing for the user, because the back button will do different things depending on where it is located on screen. But if you found that more intuitive, I'll change it.
Thank you both for your comments.

traunts commented:
Had a moment of inspiration: you don't need to hide the tabs, just have the `q/e` swap hot keys only cycle through the current group. That is, when in equipment selection mode, highlight the weapons / explosives / tools etc. tabs and have the selection loop over these. Choosing a tab outside of this group would be the same as going 'back'.

garethyr commented:
Oooh, very good idea Stewie. Change tab colour for "Active" tabs and have q/e only swap between them. That seems like a very solid middleground to me!

Flameport commented:
this sounds good

garethyr commented:
Note: this was taken over and completed [here](https://github.com/cortex-command-community/Cortex-Command-Community-Project-Source/pull/372).
